### PR TITLE
Fixes structs to handle non-text struct fields.

### LINKED ISSUE #355: Prevent structs with non-text struct keys from being created
So, this is odd:
```
PartiQL> { 1: 42 }
   | 
===' 
{
  'null': 42
}
--- 
OK!
```
1. An integer as a field key?
1. Why does it show null there instead of 1?

Additionally...

```
val strangeStruct = eval("{1: 2}")
val strangeField = strangeStruct.first().name!!
println(strangeField.type)
println(strangeField.toString())
```

outputs:
```
INT
1
```
Lastly:

```
PartiQL> { 1: 42 }[1]
   | 
org.partiql.lang.eval.EvaluationException: Expected text: 1
	<UNKNOWN>: at line 1, column 11: <UNKNOWN>
```

So it would appear that we allow structs to be created with non-text field names but then once created the resulting value is more-or-less unusable.  

I think we need to add an evaluation-time check that will raise an error if an attempt is made to construct a struct with a non-text, null or missing key.

### PR DESCRIPTION
**Pull request by abhikuhikar**
* Resolves #355 

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.

### CODE REVIEW

File: lang/src/org/partiql/lang/eval/EvaluatingCompiler.kt [hunk 1]
```diff
@@ -868,11 +868,30 @@ internal class EvaluatingCompiler(
 
         val fieldThunks = fields.map {
             val (nameExpr, valueExpr) = it
+            if (nameExpr is Literal && !nameExpr.ionValue.isText) {
+                // Compile time error. Fail before the StructFieldThunks are evaluated.
+                err("Found struct field to be of type ${nameExpr.ionValue.type}",
+                    ErrorCode.EVALUATOR_NON_TEXT_STRUCT_FIELD,
```
Comment (abhikuhikar) [inline lang/src/org/partiql/lang/eval/EvaluatingCompiler.kt:874]: Throw `SEMANTIC` exception here

File: lang/src/org/partiql/lang/errors/ErrorCode.kt [hunk 2]
```diff
@@ -649,6 +649,11 @@ enum class ErrorCode(private val category: ErrorCategory,
             "LIMIT value must be an integer but found ${errorContext.getProperty(Property.ACTUAL_TYPE)}}"
     },
 
+    EVALUATOR_NON_TEXT_STRUCT_FIELD (
```
Comment (alancai98) [inline lang/src/org/partiql/lang/errors/ErrorCode.kt:652]: nit: extra space before left paren

File: lang/src/org/partiql/lang/errors/ErrorCode.kt [hunk 3]
```diff
@@ -649,6 +649,11 @@ enum class ErrorCode(private val category: ErrorCategory,
...
+        ErrorCategory.EVALUATOR,
+        LOCATION,
+        "Struct field should be text."),
```
Comment (alancai98) [inline lang/src/org/partiql/lang/errors/ErrorCode.kt:655]: Is it more accurate to say struct field *names*/*keys* rather than struct field? A struct field has a name/key (restricted to text) and a field (could be any value).

File: lang/src/org/partiql/lang/eval/EvaluatingCompiler.kt [hunk 4]
```diff
@@ -868,11 +868,30 @@ internal class EvaluatingCompiler(
 
         val fieldThunks = fields.map {
             val (nameExpr, valueExpr) = it
+            if (nameExpr is Literal && !nameExpr.ionValue.isText) {
+                // Compile time error. Fail before the StructFieldThunks are evaluated.
+                err("Found struct field to be of type ${nameExpr.ionValue.type}",
```
Comment (alancai98) [inline lang/src/org/partiql/lang/eval/EvaluatingCompiler.kt]: nit: similar to comment in `ErrorCode.kt`, error message should say struct field *name*/*key*. Same w/ on L887

File: lang/test/org/partiql/lang/eval/EvaluatingCompilerExceptionsTest.kt [hunk 5]
```diff
@@ -242,4 +242,37 @@ class EvaluatingCompilerExceptionsTest : EvaluatorTestBase() {
...
+        )
+
+    @Test
+    fun variableReferenceToIntAsNonTextStructField() =
+        checkInputThrowingEvaluationException(
+            "SELECT {a : 2} FROM {'a' : 1}",
+            ErrorCode.EVALUATOR_NON_TEXT_STRUCT_FIELD,
+            mapOf(
+                Property.LINE_NUMBER to 1L,
+                Property.COLUMN_NUMBER to 8L
```
Comment (alancai98) [inline lang/test/org/partiql/lang/eval/EvaluatingCompilerExceptionsTest.kt:253]: There's some slight inconsistency with the error location. If the non-text struct field name is caught at compile time (as in the previous two tests), the error points to the field name. Whereas, in this case it points to the start of the struct constructor (L8). I think both cases should be consistent with the error pointing to the problematic struct field name.

File: lang/test/org/partiql/lang/eval/EvaluatingCompilerExceptionsTest.kt [hunk 6]
```diff
@@ -242,4 +242,37 @@ class EvaluatingCompilerExceptionsTest : EvaluatorTestBase() {
...
+    fun timestampAsNonTextStructField() =
+        checkInputThrowingEvaluationException(
+            "{`2007-02-23T12:14Z` : 3}",
+            ErrorCode.EVALUATOR_NON_TEXT_STRUCT_FIELD,
+            mapOf(
+                Property.LINE_NUMBER to 1L,
+                Property.COLUMN_NUMBER to 2L
+            )
+        )
+
```
Comment (alancai98) [inline lang/test/org/partiql/lang/eval/EvaluatingCompilerExceptionsTest.kt]: May want to also add some tests with `NULL` and `MISSING` as the struct field name.

File: lang/src/org/partiql/lang/errors/ErrorCode.kt (hunk 3 above)
Comment (alancai98) [reply lang/src/org/partiql/lang/errors/ErrorCode.kt:655]: Also, I don't think the `messagePrefix` argument of `ErrorCode` should be used. It can lead to a misconfigured error message:

```
PartiQL> { 1 : 2 };
org.partiql.lang.eval.EvaluationException: Found struct field to be of type INT
	Evaluator Error: at line 1, column 3: Struct field should be text., <UNKNOWN> : <UNKNOWN>

	at org.partiql.lang.eval.ExceptionsKt.err(Exceptions.kt:51)
        ...
```

Overriding `getErrorMessage` and leaving `messagePrefix` as an empty string should get rid of the extra `, <UNKNOWN> : <UNKNOWN>` at the end.

**Review by alancai98 — CHANGES_REQUESTED** [review]
I have some minor suggestions to adjust the error message and error location. Also may need a couple more tests with different struct field name types.

File: lang/test/org/partiql/lang/eval/EvaluatingCompilerExceptionsTest.kt (hunk 5 above)
Comment (abhikuhikar) [reply lang/test/org/partiql/lang/eval/EvaluatingCompilerExceptionsTest.kt]: `metas` are not available with the `ExprValue` once it is evaluated.

File: lang/test/org/partiql/lang/eval/ScratchPad.kt [hunk 7]
```diff
@@ -0,0 +1,8 @@
+package org.partiql.lang.eval
```
Comment (alancai98) [inline lang/test/org/partiql/lang/eval/ScratchPad.kt:1]: File not needed

**Review by alancai98 — COMMENTED** [review]
Looks good. Only have a couple small nits and an unused file.

**Comment by alancai98** [comment]
For future tracking: when the evaluator is to support a permissive typing mode, these fields with a non-string struct key would be ignored. From the [spec](https://partiql.org/assets/PartiQL-Specification.pdf#page=25)
>**Treatment of mistyped attribute names** It is possible that an expression a_I that computes an attribute name
results into a non-string, i.e., a value that is not a legitimate attribute name. In such cases, under the permissive mode
the attribute-value pair will be dismissed. Under the type checking mode the query will fail.
